src/google/protobuf/json/internal/parser.cc

```diff
@@ -625,6 +625,180 @@ absl::Status ParseArray(JsonLexer& lex, Field<Traits> field, Msg<Traits>& msg) {
+      Traits::SetUInt32(key_field, entry, n);
+      break;
+    }
+    case FieldDescriptor::TYPE_BOOL: {
+      if (key.value == "true") {
+        Traits::SetBool(key_field, entry, true);
+      } else if (key.value == "false") {
+        Traits::SetBool(key_field, entry, false);
+      } else {
+        return key.loc.Invalid(absl::StrFormat("expected bool string, got '%s'",
+                                               key.value.AsView()));
+      }
+      break;
+    }
+    case FieldDescriptor::TYPE_STRING: {
+      Traits::SetString(key_field, entry, std::move(key.value.ToString()));
+      break;
+    }
+    default:
+      return key.loc.Invalid("unsupported map key type");
+  }
+  return absl::OkStatus();
+}
+
+// Parses map entry when value is an enum name (string) and ignoring unknown
+// fields flag is set.
+// Specification says that in this case, if the enum name is unknown, we should
+// skip the map entry.
+// For example {"key1": "UNKNOWN_ENUM_VALUE"} should parse into an empty map if
+// the ignore_unknown_fields is set.
+//
+// The default implementation in ParseMapEntry will first allocate a new map
+// entry message (by calling Traits::NewMsg) and then consume the value from
+// the lexer.
+//
+// In this special case, we must do it the other way around:
+// - first consume the value from the lexer,
+// - check that this is a known enum value,
+// - and only then proceed with allocating a new map entry message.
+//
+// Note that in both cases the key is already consumed and passed in through
+// 'LocationWith<MaybeOwnedString>& key' param.
+template <typename Traits>
+absl::Status ParseMapEntryWithEnumStringValueWhenIgnoringUnknownFields(
+    JsonLexer& lex, Field<Traits> map_field, Msg<Traits>& parent_msg,
+    LocationWith<MaybeOwnedString>& key) {
+  // Parse the enum value from string, advancing the lexer.
+  absl::StatusOr<absl::optional<int32_t>> enum_value;
+  (void)Traits::WithFieldType(
+      map_field, [&lex, &enum_value](const Desc<Traits>& map_entry_desc) {
+        enum_value = ParseEnum<Traits>(lex, Traits::ValueField(map_entry_desc));
+        return absl::OkStatus();
+      });
+  RETURN_IF_ERROR(enum_value.status());
+
+  // If enum value is unknown, we'll stop here and avoid allocating a new map
+  // entry message.
+  if (!enum_value->has_value()) {
+    return absl::OkStatus();
+  }
+
+  // If the enum value is known, output the map entry with already parsed
+  // value.
+  return Traits::NewMsg(
+      map_field, parent_msg,
+      [&](const Desc<Traits>& type, Msg<Traits>& entry) -> absl::Status {
+        RETURN_IF_ERROR(ParseMapKey<Traits>(lex.options(), type, entry, key));
+
+        // This is different from the default implementation in 'ParseMapEntry'
+        // since we already parsed this value (we must not attempt to advance
+        // the lexer to parse the value here again).
+        Traits::SetEnum(Traits::ValueField(type), entry,
+                        enum_value->value_or(0));
```

Comment:
Nit: this should just be `*enum_value` (or `enum_value->value()`), you already checked that has_value is true above so it will be guaranteed to succeed. 

It's academic in this case because the 0 value would never be used, but under proto2 semantics (where enums are closed, which is the case you are handling here) the default value for an enum is whatever is listed first in the definition, which isn't necessarily 0 value for the enum; there could either be a 0 value that is not listed first (then 0 is a legal value but not the default) or no listed 0 value at all (then its not a legal value at all).

Only under proto3 / open enums is it mandatory that the enum has a 0 value and that it is listed first (exactly to allow every enum to have a 0 default in proto3).